server/src/main/java/com/broadcom/lsp/cobol/service/CopybookServiceImpl.java

```diff
@@ -47,77 +47,110 @@
+      @Named("CACHE-MAX-SIZE") int cacheSize,
+      @Named("CACHE-DURATION") int duration,
+      @Named("CACHE-TIME-UNIT") String timeUnitName) {
     this.dataBus = dataBus;
     this.settingsService = settingsService;
     this.files = files;
-
-    dataBus.subscribe(REQUIRED_COPYBOOK_EVENT, this);
+    copybookCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(duration, TimeUnit.valueOf(timeUnitName))
+        .maximumSize(cacheSize)
+        .build();
     dataBus.subscribe(ANALYSIS_FINISHED_EVENT, this);
   }
 
   @Override
-  public void invalidateURICache() {
-    copybookPaths.clear();
+  public void invalidateCache() {
     copybooksForDownloading.clear();
+    copybookCache.invalidateAll();
   }
 
   /**
-   * Depends on DataEvent type it will be handled with appropriate handler: {@link
-   * #handleRequiredCopybookEvent} or {@link #handleAnalysisFinishedEvent}.
+   * Retrieve and return the copybook by its name. Copybook may cached to limit interactions with
+   * the file system.
    *
-   * @param event the instance of {@link RequiredCopybookEvent} or {@link AnalysisFinishedEvent}
+   * Resolving works in synchronous way. Resolutions with different copybook names will not block each other.
+   *
+   * @param copybookName - the name of the copybook to be retrieved
+   * @param documentUri - the currently processing document that contains the copy statement
+   * @param copybookProcessingMode - text document synchronization type
+   * @return a CopybookModel that contains copybook name, its URI and the content
    */
-  @Override
-  public void observerCallback(DataEvent event) {
-    if (event instanceof RequiredCopybookEvent) {
-      handleRequiredCopybookEvent((RequiredCopybookEvent) event);
-    } else if (event instanceof AnalysisFinishedEvent) {
-      handleAnalysisFinishedEvent((AnalysisFinishedEvent) event);
+  @Nullable
+  public CopybookModel resolve(
+      @Nonnull String copybookName,
+      @Nonnull String documentUri,
+      @Nonnull CopybookProcessingMode copybookProcessingMode) {
+    if (!copybookProcessingMode.analyze) {
+      return new CopybookModel(copybookName, null, null);
+    }
+    boolean firstRequest = false;
+    CompletableFuture<CopybookModel> futureModel;
+    synchronized (this) {
+      futureModel = copybookCache.getIfPresent(copybookName);
+      if (futureModel == null) {
+        futureModel = new CompletableFuture<>();
+        firstRequest = true;
+        copybookCache.put(copybookName, futureModel);
+      }
+    }
+    try {
+      if (firstRequest) {
+        CopybookModel copybookModel = resolveSync(copybookName, documentUri, copybookProcessingMode);
+        futureModel.complete(copybookModel);
+        return copybookModel;
+      } else {
+        return futureModel.get(30, TimeUnit.SECONDS);
+      }
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      LOG.error("Can't resolve copybook '{}'.", copybookName, e);
+      return null;
```

Comment:
is it possible to return return new CopybookModel(copybookName, null, null); instead of null as like other places? 
Otherwise Optional<CopybookModel>

Reply:
I agree, returning optional would be a good idea here

Reply:
Yes, makes sense.